src/worker.js

```diff
@@ -134,12 +171,67 @@ worker.cleanupAndExit = function(code) {
   }
 }
 
+worker.tryCleanup = function() {
+  var _exit = function() {
+    worker.exit();
+  }
+
+  var _abort = function() {
+    if (!worker.abortListeners.length) {
+      worker.abortListeners = [];
+    }
+  }
+
+  if (worker.abortListeners.length) {
+    const promises = worker.abortListeners.map(listener => listener());
+    let timerId;
+    const timeoutPromise = new Promise((_resolve, reject) => {
+      timerId = setTimeout(reject, worker.abortListenerTimeout);
+    });
+
+    // Once a promise settles we need to clear the timeout to prevet fulfulling the promise twice 
+    const settlePromise = Promise.all(promises).then(function() {
+      clearTimeout(timerId);
+      _abort();
+    }, function() {
+      clearTimeout(timerId);
+      _exit();
+    });
+
+
+    return Promise.all([
+      settlePromise,
+      timeoutPromise
+    ]);
+  }
+  // if there are no listeners just reject in a promise and let the worker cleanup start
+  return new Promise(function(_resolve, reject) { reject(); });
```

Comment:
Can you explain this to me? It feels odd triggering a rejection when there is no error occurring, and it is also a bit odd to leave err `undefined`, i.e. should be something like `reject(new Error('...'))`

Reply:
Good spot and yes we should be throwing an instance of `Error` here. I have updated this and also removed the `worker.kill()` invocation of the `tryCleanup` method as it will be handled through the `workerHandler` `cleanup` function.


With these changes I now get the following output from your isolated test implementation


```
start
TimeoutError: Promise timed out after 50 ms
{ result: 7 }
pool1 workers: 0
done
```


I have also reverted the `afterEach` hook back to `after` see https://github.com/josdejong/workerpool/pull/448#discussion_r1786896894

Reply:
I wanted to move this logic into `tryCleanup` where we send the message to the main thread the cleanup result. With the case of `worker.abortListeners.length === 1` returning a promise tracking the result of the handlers or a timeout promise. not returning a promise in the case of `worker.abortListeners.length === 0` diverges the return value of the function depending on the length of the `abortListeners` array.

Reply:
I did a bit of refactoring and we could refactor `tryCleanup` in the following way

```js
worker.cleanup = function(requestId) {
  var _exit = function() {
    worker.exit();
  }

  var _abort = function() {
    if (!worker.abortListeners.length) {
      worker.abortListeners = [];
    }
  }

  if (worker.abortListeners.length) {
    const promises = worker.abortListeners.map(listener => listener());
    let timerId;
    const timeoutPromise = new Promise((_resolve, reject) => {
      timerId = setTimeout(function () { 
        reject(new Error('Timeout occured waiting for abort handler, killing worker'));
      }, worker.abortListenerTimeout);
    });

    // Once a promise settles we need to clear the timeout to prevet fulfulling the promise twice 
    const settlePromise = Promise.all(promises).then(function() {
      clearTimeout(timerId);
      _abort();
    }, function() {
      clearTimeout(timerId);
      _exit();
    });


    return Promise.all([
      settlePromise,
      timeoutPromise
    ]).then(function() {
      worker.send({
        id: requestId,
        method: CLEANUP_METHOD_ID,
        error: null,
      });
    }, function(err) {
      worker.send({
        id: requestId,
        method: CLEANUP_METHOD_ID,
        error: err ? convertError(err) : null,
      });
    });
  }

  worker.send({
    id: requestId,
    method: CLEANUP_METHOD_ID,
    error: convertError(new Error('Worker terminating')),
  });
  return;
}
```

I am not loving this as there is one branch where the method returns a `Promise` and another where the function is retuning `void` but we still need to `await` the call to `cleanup` as if `abortListeners` has a handler we must await it. 

I like the previous implementation better as it allows a promise to be return in either branch which to me is a better implementation, but curious what your thoughts are.

Reply:
Hm, yeah 🤔.

I think the confusion on my side originates from the cleanup function having three different cases, but returning only two cases: 

1. there where abort listener(s) and they succesfully aborted, so no need to terminate the worker
2. there where abort listener(s) but something went wrong, so we need to terminate the worker 
3. there where no abort listeners, so we need to terminate the worker

Now, returning two cases works fine because in case (2) and (3) we need to take the same action: terminate the worker. We could implement something to report back these three different cases to the WorkerHandler to make it explicit, or we can leave it as it is right now and clarify with some extra comments.

Idea:

- Rename the function `tryCleanup` to `cleanup`
- Rename function `cleanupAndExit` to `terminateAndExit` to prevent any confusion with `cleanup`.
- Move all logic inside `cleanup` (like your last proposal), and add a comment explaining the three cases. Having it contained in one function with a bit of extra explanation may make it easier to understand and refactor in the future.
- Change the function `cleanup` to _always_ return `Promise<void>`
- Change function `cleanupAndExit` to return `Promise<void>` too (it contains a promise but no return right now)
- Change the error message "Cleanup failed, exiting worker" into "Worker terminating", that is more clear to me.
- Change `worker.on('message', function (request) {...}` into `worker.on('message', async function (request) { ...}`, remove the two `return` cases, replace them with `await`.
- Extra: reduce nesting in function `cleanup` by exiting early in case of `worker.abortListeners.length === 0`, moving this `if` to the top of the function and inverting it. That will also make `_abort` redundant I think, and you could inline `_exit` since it is a one liner which is only used once.

Reply:
I think we ultimately will need to implement some piece of context to correspond to the result of task execution. If we do this then there are many solutions to cleanup of worker handlers which are found to be in some terminal execution state and need to be killed and restarted. 

I was hesitant to implement this with this pr as I was unsure if we wanted to go down this route. I think for now we should go with your outline as changes to how workers are terminated could have adverse effects on up stream projects.

I think the approach you outline as the initial solution makes sense but I have a question regarding

> Change worker.on('message', function (request) {...} into worker.on('message', async function (request) { ...}, remove the two return cases, replace them with await

Would we effect up stream projects with the use of `async / await` in code that is released? If so we might want to keep with the current option of returning a promise. So we might need to keep the pattern of returning a promise in both cases of either `successful` or `error` the handler execution may result in.

Reply:
> The main reason to make the callback async is to prevent vague (future) issues when there are multiple async actions inside the callback that we would expect to run after each other. Using await in the code makes it clear that an async function is being executed.



Can't we achieve this with the current pattern of returning the `Promise` from `cleanup` in the `worker` message handler? 

I think if the user wants to implement multiple async actions from within the callback then they may declare the callback function to be `async`. WDYT?

Reply:
Thanks for the updates, I think the function `cleanup` now contains enough explanation for my futher me to understand what's going on 😉. 

> Can't we achieve this with the current pattern of returning the `Promise` from `cleanup` in the `worker` message handler?

Yes, that is also fine with me 👍. I like the `async` and return `Promise<void>` slightly better but if you prefer this approach I'm fine with that.

Last remark: I think the following line is now sort of redundant, and can be reduced to just `return`, or returning a resolved promise. Then, the function consistently returns `Promise<void>` in all three cases. Does that make sense?

```js
return new Promise(function(reject) { reject(new Error("Worker terminating"));
```
The custom `Promise` of workerpool doesn't have a `Promse.resolve`, so this will look like `return new Promise(resolve => resolve());`